compiler/rustc_errors/src/emitter.rs

```diff
@@ -1653,6 +1661,12 @@ impl EmitterWriter {
             }
         }
 
+        if self.track_diagnostics && let Some(tracked) = emitted_at {
```

Comment:
Seems weird to have two places that decide this - can we move the `track_diagnostics` check wherever `emitted_at` is created and pass in None if it's false? Then we can remove this check.

Reply:
I have changed to pass in Some/None where `emit_messages_default` is called, but I'm not sure how to do it where `emitted_at` is created, because that's in `Diagnostic::new_with_code`, and I don't know how to access/check flags there. What's the best approach here?

Reply:
Not sure, sorry. If this is hard it's ok to keep the code as is.